fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,470 @@
+
+    See https://gist.github.com/saltiniroberto/9ee53d29c33878d79417abb2b4468c20 for an explanation of why this is
+    required.
+    """
+    return Gwei(ceil_div(int(estimate * (1000 + COMMITTEE_WEIGHT_ESTIMATION_ADJUSTMENT_FACTOR)), 1000))
+```
+
+#### `get_committee_weight_between_slots`
+
+```python
+def get_committee_weight_between_slots(state: BeaconState, start_slot: Slot, end_slot: Slot) -> Gwei:
+    """
+    Returns the total weight of committees between ``start_slot`` and ``end_slot`` (inclusive of both).
+    """
+    total_active_balance = get_total_active_balance(state)
+
+    start_epoch = compute_epoch_at_slot(start_slot)
+    end_epoch = compute_epoch_at_slot(end_slot)
+
+    if start_slot > end_slot:
+        return Gwei(0)
+
+    # If an entire epoch is covered by the range, return the total active balance
+    if is_full_validator_set_covered(start_slot, end_slot):
+        return total_active_balance
+
+    if start_epoch == end_epoch:
+        return Gwei(ceil_div((end_slot - start_slot + 1) * int(total_active_balance), SLOTS_PER_EPOCH))
+    else:
+        # A range that spans an epoch boundary, but does not span any full epoch
+        # needs pro-rata calculation
+
+        # See https://gist.github.com/saltiniroberto/9ee53d29c33878d79417abb2b4468c20
+        # for an explanation of the formula used below.
+
+        # First, calculate the number of committees in the end epoch
+        num_slots_in_end_epoch = int(compute_slots_since_epoch_start(end_slot) + 1)
+        # Next, calculate the number of slots remaining in the end epoch
+        remaining_slots_in_end_epoch = int(SLOTS_PER_EPOCH - num_slots_in_end_epoch)
+        # Then, calculate the number of slots in the start epoch
+        num_slots_in_start_epoch = int(SLOTS_PER_EPOCH - compute_slots_since_epoch_start(start_slot))
+
+        end_epoch_weight_mul_by_slots_per_epoch = num_slots_in_end_epoch * int(total_active_balance)
+        start_epoch_weight_mul_by_slots_per_epoch = ceil_div(
+            num_slots_in_start_epoch * remaining_slots_in_end_epoch * int(total_active_balance),
+            SLOTS_PER_EPOCH
+        )
+
+        # Each committee from the end epoch only contributes a pro-rated weight
+        return adjust_committee_weight_estimate_to_ensure_safety(
+            Gwei(ceil_div(
+                start_epoch_weight_mul_by_slots_per_epoch + end_epoch_weight_mul_by_slots_per_epoch,
+                SLOTS_PER_EPOCH
+            ))
+        )
+```
+
+#### `is_one_confirmed`
+
+```python
+def is_one_confirmed(store: Store, block_root: Root) -> bool:
+    current_slot = get_current_slot(store)
+    block = store.blocks[block_root]
+    parent_block = store.blocks[block.parent_root]
+    support = int(get_weight(store, block_root))
+    justified_state = store.checkpoint_states[store.justified_checkpoint]
+    maximum_support = int(
+        get_committee_weight_between_slots(justified_state, Slot(parent_block.slot + 1), Slot(current_slot - 1))
```

Comment:
Do we assume that we run the protocol at the beginning of each slot, when the block for the current slot is not proposed? Otherwise, for the block proposed in the current slot (with its parent proposed in the previous slot),  Slot(parent_block.slot + 1) > Slot(current_slot - 1).

Reply:
The protocol is run at the beginning of each slot regardless of whether we propose or not a block in that slot. Also, we do not run the protocol on blocks for the current slots regardless.